File: 3475.cpp
#include <cstdio>
#include <algorithm>
#include <cstring>

using namespace std;
const int C = 50000,N = 1000000 - C;
int n = 0,tail = 1 ,head = 1,c,t;
int s[N],q[C];
bool empty = 1,dir = 1;

int next(int x)
{
	if (dir) return (x % c) + 1;
	else return (x+c-2) % c + 1;
}
int front(int x)
{
	if (dir) return (x + c - 2) % c + 1;
	else return (x % c) + 1;
}

void push()
{
	int num,ne;
	scanf("%d",&num);
	ne = next(head);
	if (ne == tail)
	{
		n ++;
		s[n] = q[tail];
		tail = next(tail);
	}
	if (empty) empty = 0;
	else head = ne;
	q[head] = num;
}

void pop()
{
	if (empty) 
	{	
		puts("Error: the stack is empty!");
		return;
	}
	printf("%d\n",q[head]);
	if (tail == head) empty = 1;
	else 
	{
		head = front(head);
		if (n > 0)
		{
			tail = front (tail);
			q[tail] = s[n];
			n --;
		}
	}
}

void reserse()
{
	int tmp;
	if (next(head) == tail)
	{
		dir = ! dir;
		tmp = head;
		head = tail;
		tail = tmp;
	} else printf("Error: less than %d elements in the stack!\n", c);
}

int main()
{
	freopen("3475.in","r",stdin);
	
	scanf("%d",&c);
	do
	{
		scanf("%d",&t);
		if (t==1) push();
		if (t==2) pop();
		if (t==3) reserse();
	} while (t!=0);
	
	return 0;
}

File: OI/纪中刘远哲/2015/8.8/3475.pas
const	maxc=50000;maxn=1000000-maxc;
var	n,c,t,tail,head,num:longint;
	s:array[1..maxn] of longint;
	q:array[1..maxc] of longint;
	dir,empty:boolean;
function next(x:longint):longint;
begin
	if dir then next:=(x mod c) + 1
	else next:=(x+c-2) mod c + 1;
end;
function front(x:longint):longint;
begin
	if dir then front:=(x+c-2) mod c +1
	else front:=(x mod c)+1;
end;
procedure push;
begin
	read(num);
	if (next(head) = tail) then begin
		inc(n);
		s[n]:=q[tail];
		tail:=next(tail);
	end;
	if empty then 
		empty:=false
	else	head:=next(head);
	q[head]:=num;
end;
procedure pop;
begin
	if empty then begin
		writeln('Error: the stack is empty!');
		exit;
	end;
	writeln(q[head]);
	if tail=head then empty:=true
	else begin
		head:=front(head);
		if n>0 then begin
			tail:=front(tail);
			q[tail]:=s[n];
			dec(n);
		end;
	end;
end;
procedure reserse;
var	tmp:longint;
begin
	if next(head) = tail then begin
		dir:=not dir;
		tmp:=head;
		head:=tail;
		tail:=tmp;
	end else writeln('Error: less than ',c,' elements in the stack!');
end;
begin
	assign(input,'3475.in');reset(input);

	readln(c);
	n:=0;
	tail:=1;
	head:=1;
	empty:=true;
	dir:=true;
	repeat
		read(t);
		case t of 
			1:push;
			2:pop;
			3:reserse;
		end;
	until t=0;
end.
